edgelet/aziot-edged/src/image_gc.rs

```diff
@@ -0,0 +1,302 @@
+                    edgelet_docker::Error::GetImageHash(),
+                )
+            })?
+            .to_string(),
+        Err(e) => {
+            log::error!("Could not get list of docker images: {}", e);
+            EdgedError::from_err("Could not get list of docker images: {}", e).to_string()
+        }
+    };
+
+    Ok(bootstrap_image_id)
+}
+
+async fn garbage_collector(
+    runtime: &edgelet_docker::DockerModuleRuntime<http_common::Connector>,
+    image_use_data: ImagePruneData,
+    bootstrap_image_id: Option<String>,
+    edge_agent_bootstrap: String,
+) -> Result<(), EdgedError> {
+    log::info!("Image Garbage Collection starting daily run");
+
+    let bootstrap_img_id = match bootstrap_image_id {
+        Some(id) => id,
+        None => {
+            let id = match get_bootstrap_image_id(runtime, edge_agent_bootstrap).await {
+                Ok(img_id) => img_id,
+                Err(e) => {
+                    log::error!("Could not get list of docker images: {}", e);
+                    EdgedError::from_err("Could not get list of docker images: {}", e).to_string()
+                }
+            };
+
+            id
+        }
+    };
+
+    // track images associated with extant containers
+    let in_use_image_ids = match ModuleRuntime::list_with_details(runtime).await {
+        Ok(modules) => {
+            let mut image_ids: HashSet<String> = HashSet::new();
+            image_ids.insert(bootstrap_img_id); // bootstrap edge agent image should never be deleted
+
+            for module in modules {
+                // this is effectively the result of calling GET on /containers/json
+                // image_hash() is created from ImageId (ImageId is filled in by GET /containers/json)
+                let id = edgelet_core::Module::config(&module.0)
+                    .image_hash()
+                    .ok_or_else(|| {
+                        EdgedError::from_err(
+                            "error getting image id for running container",
+                            edgelet_docker::Error::GetImageHash(),
+                        )
+                    })?;
+                image_ids.insert(id.to_string());
+            }
+            image_ids
+        }
+        Err(err) => {
+            return Err(EdgedError::new(format!(
+                "Error in image auto-pruning task. Cannot get running modules. Skipping image auto pruning. {}",
+                err
+            )));
+        }
+    };
+
+    let image_map = image_use_data
+        .prune_images_from_file(in_use_image_ids)
+        .map_err(|e| {
+            EdgedError::from_err(
+                "Image garbage collection failed to prune images from file.",
+                e,
+            )
+        })?;
+
+    // delete images
+    for key in image_map.keys() {
+        if let Err(e) = ModuleRegistry::remove(runtime, key).await {
+            log::error!("Could not delete image {} : {}", key, e);
```

Comment:
If this error occurs, the image may never get deleted, right? Consider adding the image back into the image usage file so that deletion is attempted again on the next iteration.

Reply:
It was considered.

Docker error isn't on us. Spewing this in the error logs will draw attention to it.

You pointing this out too, makes me a tad more ambivalent though.

Maybe the images do need to get added back....

Reply:
I could go either way on this. If we've already confirmed that there's no container for this image (the image wouldn't be in this list otherwise), then it is unlikely docker will fail to delete it. Of course, Docker could go down, or a container could get created after the list was created but before we try to delete the image. But I think the change of this happening is remote, so I don't know how elaborate our handling needs to be.

@gauravIoTEdge how hard is it to add the image back to the list in the current impl?

Reply:
`I think the change of this happening is remote, so I don't know how elaborate our handling needs to be`

^^ that was my thinking too.

`or a container could get created after the list was created but before we try to delete the image`

^^ I'd considered that too. Which is when the realization dawned on me that this is analysis-paralysis territory.

Ultimately, I chose to not add the images back to adopt a wait-and-watch approach --- let's wait to see if this is actually a problem and see if users complain. Adding the images back in (with the current implementation) is a non-trivial amount of work because it involves one of:

1. Adding a public API to write directly to the state file, OR
2. Keeping the old state from being clobbered by the new state for long enough till deletion is confirmed (sub-optimal and not as good as # 1 because it will have information on images that may have been successfully deleted as well)